### ISSUE TITLE
Group service instances by required set of tags: pattern?

### ISSUE THREAD

**Issue by djenriquez** [hashicorp/consul-template#403]
Related to #398 and #391, is there an easy way to group services by their actual instance tags?

Consul is grouping everything by Service Name and the .Tags.Contains is ran against the service level, as you know. However, how would we filter based on a set of tags within a Service?

For example: service “Foo” looks like this in Consul:
```
{
	Service: {
       		Name: Foo,
		Tags: [
			“dev”,
			“test”,
			“prod”,
			“ab”,
			“proxied"
		],
		Service_instances:[
			{
				ID: 1,
				Name: Foo,
				Tags: [
					“dev”
				]
			},
			{
				ID: 2,
				Name: Foo,
				Tags: [
					“dev”,
					“proxied”
				]
			},
			{
				ID: 3,
				Name: Foo,
				Tags: [
					“test”,
					“ab"
				]
			},
			{
				ID: 4,
				Name: Foo,
				Tags: [
					“prod”,
					“ab”,
					“proxied"
				]
			},
		]
	}
}
```
In consul-template, If I’m in the dev environment and I do something like:

```
{{$app_env := “prod”}}
.
{{range services}}
{{if and (.Tags.Contains $app_env) (.Tags.Contains “proxied”)}}
//this will process all service instances, but I only want it to process service 4
{{end}}
{{end}}
```

Also, what about the situation where we want to filter out service instances by X number of tags? Is there a certain pattern that would handle this?

**Comment by sethvargo** [comment]
Hi @djenriquez 

Thank you for opening an issue. I'm sorry, but I re-read the issue 3x and I truly do not understand what you are asking. This feels like a brain dump instead of an actual question, and therefore I am unsure how to answer it :frowning:. Please keep in mind that we are not inside your infrastructure and have no familiarity with your setup.

I think you're misunderstanding how services are represented in Consul. Consul Template does not query for node tags, only service tags. One of the key components of service discovery is that you do not worry about these little pieces. "prod" and "test" should likely be their own data centers or a different service entirely.

Additionally, Consul's API does not return the data in which you've described it. The services catalog endpoint looks like this:

```javascript
[
  {
    "Node": "node-10-0-3-152",
    "Address": "10.0.3.152",
    "ServiceID": "a",
    "ServiceName": "a",
    "ServiceTags": null,
    "ServiceAddress": "",
    "ServicePort": 8588
  },
  {
    "Node": "node-10-0-1-28",
    "Address": "10.0.1.28",
    "ServiceID": "a",
    "ServiceName": "a",
    "ServiceTags": null,
    "ServiceAddress": "",
    "ServicePort": 8588
  }
]
```

Keep in mind that you are also querying Consul's *catalog* by using the `services` api. The use cases for doing this are actually quite rare. Usually you want to use the `service` api which will query the *health* endpoints for healthy services.

I think you may need to re-think or re-archiect the way you are storing data in Consul. It sounds like you are severely abusing tags. This is not only not-optimized (Consul doesn't optimize for tags), but it's not a recommended pattern and is not really supported to do complex querying by tags. If your services are so very different, they should either be completely different services or live in a different data center.

We like to keep the issue tracker reserved for bugs and feature requests and ask that usage-related questions be posted to the [Consul mailing list](https://groups.google.com/forum/#!forum/consul-tool). There is a much larger community there and you will get a faster response than GitHub issues.

**Comment by djenriquez** [comment]
Hi @sethvargo! Haha i'm sorry, i didnt mean to overcomplicate things. I do know that the data structure i posted is not what consul returns, it was supposed to be a simplification; just the abstracted service picture.

Basically what I was trying to explain was, for example, we have 4 instances of the Service "Foo". By instance, I mean an actual instance of the service, not a node or server. They are the exact same service in all 4 instances. 

We then loop through all the Services by name using `{{range services}}`. Then for each service we set up a template to create an NGINX config based on the tags each service had. I thought this was what the tagging was built for?

```
{{range services}}
{{range service .Name}}
server{
    location / {
{{if .Tags.Contains "dev"}}
         proxy_pass http://{{ .Name }}_dev
{{end}}
    }
}
{{end}}
```

So this will create
```
server{
     location / {
         proxy_pass http://Foo_dev
     }
}
```

We then take this a step further by importing an environment variable `$app_env` with `{{$app_env := (env "APP_ENV")}}`. So now, I can reuse the same template in dev,test,staging, and prod. 

```
{{$app_env := (env "APP_ENV")}}
{{range services}}
{{range service .Name}}
server{
    location / {
{{if .Tags.Contains $app_env}}
         proxy_pass http://{{ .Name }}_{{$app_env}}
{{end}}
    }
}
{{end}}
```

However, for each environment, I only want the appropriate service to be loaded into the NGINX config. So, with the abstracted service picture example from the original post, the template above will create:

```
server{
     location / {
         proxy_pass http://Foo_dev
     }
}
server{
     location / {
         proxy_pass http://Foo_dev
     }
}
```

It creates it twice because there are two services tagged `dev`. This is solved with #391 indexing.

```
{{$app_env := (env "APP_ENV")}}
{{range services}}
{{range $i, $s := service .Name}}
{{if eq $i 0}}
server{
    location / {
{{if .Tags.Contains $app_env}}
         proxy_pass http://{{ $s.Name }}_{{$app_env}}
{{end}}
    }
}
{{end}}
{{end}}
```

But now, I only want to process it if the service contains BOTH `dev` and `proxied`. This is my issue. I'm unsure how to do this because if I don't know which index has the `proxied` tag. If i put an `.Tags.Contains`, I cannot guarantee the 0th index is valid. So the below works sometimes IF what i'm looking for is the 0th index.

```
{{$app_env := (env "APP_ENV")}}
{{range services}}
{{range $i, $s := service .Name}}
{{if and (.Tags.Contains $app_env) (.Tags.Contains "proxied)}}
{{if eq $i 0}}
server{
    location / {
{{if .Tags.Contains $app_env}}
         proxy_pass http://{{ $s.Name }}_{{$app_env}}
{{end}}
    }
}
{{end}}
{{end}}
{{end}}
```

I'll post this up in the mailing list, but I figured I would explain so that I didn't leave you in a brain-dumped state haha.

**Comment by djenriquez** [comment]
Heres the [link](https://groups.google.com/forum/#!topic/consul-tool/wpppi8eW2L8) to the mailing list question, also a different perspective to the same problem.
